### ISSUE TITLE
[Bug] autograd of posterior mean w.r.t. x seems unstable

### ISSUE THREAD

**Issue by shalijiang** [pytorch/botorch#280]
# 🐛 Bug

trying to compute the norm of gradient of posterior mean, but find different runs of autograd give different results. 
Not sure if this is a general problem of autograd or specific to the model.posterior(x).mean function.
Note the eigenvalues of the kernel matrix show that it's not ill-conditioned.  

## To reproduce

** Code snippet to reproduce **
```python

import torch
from torch import Tensor
import pickle
# from botorch.models import SingleTaskGP
from typing import Optional

import torch
from gpytorch.constraints.constraints import GreaterThan
from gpytorch.distributions.multivariate_normal import MultivariateNormal
from gpytorch.kernels.matern_kernel import MaternKernel
from gpytorch.kernels.scale_kernel import ScaleKernel
from gpytorch.likelihoods.gaussian_likelihood import (
    FixedNoiseGaussianLikelihood,
    GaussianLikelihood,
    _GaussianLikelihoodBase,
)
from gpytorch.likelihoods.likelihood import Likelihood
from gpytorch.likelihoods.noise_models import HeteroskedasticNoise
from gpytorch.means.constant_mean import ConstantMean
from gpytorch.models.exact_gp import ExactGP
from gpytorch.priors.smoothed_box_prior import SmoothedBoxPrior
from gpytorch.priors.torch_priors import GammaPrior
from torch import Tensor

from botorch.models.gpytorch import BatchedMultiOutputGPyTorchModel

MIN_INFERRED_NOISE_LEVEL = 1e-6
class SingleTaskGP(ExactGP, BatchedMultiOutputGPyTorchModel):
    r"""A single-task Exact GP model.

    A single-task exact GP using relatively strong priors on the Kernel
    hyperparameters, which work best when covariates are normalized to the unit
    cube and outcomes are standardized (zero mean, unit variance).

    This model works in batch mode (each batch having its own hyperparameters).
    When the training observations include multiple outputs, this model will use
    batching to model outputs independently.

    Use this model when you have independent output(s) and all outputs use the same
    training data. If outputs are independent and outputs have different training
    data, use the ModelListGP. When modeling correlations between outputs, use
    the MultiTaskGP.
    """

    def __init__(
        self, train_X: Tensor, train_Y: Tensor, likelihood: Optional[Likelihood] = None
    ) -> None:
        r"""A single-task Exact GP model.

        Args:
            train_X: A `n x d` or `batch_shape x n x d` (batch mode) tensor of training
                features.
            train_Y: A `n x (o)` or `batch_shape x n x (o)` (batch mode) tensor of
                training observations.
            likelihood: A likelihood. If omitted, use a standard
                GaussianLikelihood with inferred noise level.

        Example:
            >>> train_X = torch.rand(20, 2)
            >>> train_Y = torch.sin(train_X[:, 0]]) + torch.cos(train_X[:, 1])
            >>> model = SingleTaskGP(train_X, train_Y)
        """
        ard_num_dims = train_X.shape[-1]
        self._set_dimensions(train_X=train_X, train_Y=train_Y)
        train_X, train_Y, _ = multioutput_to_batch_mode_transform(
            train_X=train_X, train_Y=train_Y, num_outputs=self._num_outputs
        )
        if likelihood is None:
            noise_prior = GammaPrior(1.1, 0.05)
            noise_prior_mode = (noise_prior.concentration - 1) / noise_prior.rate
            likelihood = GaussianLikelihood(
                noise_prior=noise_prior,
                batch_shape=self._aug_batch_shape,
                noise_constraint=GreaterThan(
                    MIN_INFERRED_NOISE_LEVEL,
                    transform=None,
                    initial_value=noise_prior_mode,
                ),
            )
        else:
            self._likelihood_state_dict = deepcopy(likelihood.state_dict())
        super().__init__(train_X, train_Y, likelihood)
        self.mean_module = ConstantMean(batch_shape=self._aug_batch_shape)
        self.covar_module = ScaleKernel(
            MaternKernel(
                nu=2.5,
                ard_num_dims=ard_num_dims,
                batch_shape=self._aug_batch_shape,
                lengthscale_prior=GammaPrior(3.0, 6.0),
            ),
            batch_shape=self._aug_batch_shape,
            outputscale_prior=GammaPrior(2.0, 1),
        )
        self.to(train_X)

    def forward(self, x: Tensor) -> MultivariateNormal:
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultivariateNormal(mean_x, covar_x)

with open("../debug_autograd.dat", "rb") as f:  # unzip the attached .zip file 
    data = pickle.load(f)
model_state = data["model_state"]
X = data["X"][0]
Y = data["Y"]
x = data["x"]
# print(x, x.shape)
model = SingleTaskGP(X, Y)
model.load_state_dict(model_state)
posterior_mean = model.posterior(x).mean
# print(posterior_mean)
num_repeats = 20
grads = [0] * num_repeats

for i in range(num_repeats):
    grads[i] = torch.autograd.grad(
        posterior_mean.squeeze(),
        x,
        grad_outputs=torch.Tensor([1.0] * len(posterior_mean)),
        create_graph=True,
    )[0]
    print(f"norm: {grads[i].norm().detach():<17}, grad: {grads[i].detach().numpy()}")


# see if the kernel matrix is ill-conditioned
K = model.covar_module(X,X).evaluate().detach().numpy()
print("eigenvalues of kernel matrix:\n", np.linalg.eig(K)[0])
```

** Stack trace/error message **
```
norm: 3.165595293045044, grad: [[ 0.31403214 -0.4016987   3.1242533  -0.0075816 ]]
norm: inf              , grad: [[-5.7399116e+23  7.5462431e+23  5.5454977e+23 -3.3287596e+23]]
norm: 3.165595293045044, grad: [[ 0.31403214 -0.4016987   3.1242533  -0.0075816 ]]
norm: 3.165595293045044, grad: [[ 0.31403214 -0.4016987   3.1242533  -0.0075816 ]]
norm: 3.165595293045044, grad: [[ 0.31403214 -0.4016987   3.1242533  -0.0075816 ]]
norm: inf              , grad: [[-5.7399116e+23  7.5462431e+23  5.5454977e+23 -3.3287596e+23]]
norm: inf              , grad: [[-3.7144462e+22  4.7384191e+22 -2.0683729e+23 -2.0664616e+22]]
norm: inf              , grad: [[-5.5315715e+22  6.1239115e+22  1.7078501e+22 -2.9080216e+21]]
norm: inf              , grad: [[-6.7333276e+20  7.3049702e+20  7.0442505e+20  4.2533686e+18]]
norm: 3.165595293045044, grad: [[ 0.31403214 -0.4016987   3.1242533  -0.0075816 ]]
norm: 3.165595293045044, grad: [[ 0.31403214 -0.4016987   3.1242533  -0.0075816 ]]
norm: 3.165595293045044, grad: [[ 0.31403214 -0.4016987   3.1242533  -0.0075816 ]]
norm: 13.015168190002441, grad: [[ 1.0258255  -3.4464521  12.508536    0.02787224]]
norm: inf              , grad: [[-1.4869729e+22  1.1159501e+22 -2.7945394e+22  9.8531644e+19]]
norm: inf              , grad: [[-3.4760918e+21 -1.2014268e+21 -3.9865152e+22  2.6559295e+19]]
norm: 3.165595293045044, grad: [[ 0.31403214 -0.4016987   3.1242533  -0.0075816 ]]
norm: inf              , grad: [[-3.3196202e+23  9.3569012e+23  3.5546688e+23 -3.9142683e+23]]
norm: inf              , grad: [[-3.4760918e+21 -1.2014268e+21 -3.9865152e+22  2.6559295e+19]]
norm: inf              , grad: [[-3.6437675e+20  3.4892328e+20 -2.0767334e+21 -1.3027703e+20]]
norm: 3.165595293045044, grad: [[ 0.31403214 -0.4016987   3.1242533  -0.0075816 ]]

eigenvalues of kernel matrix:
 [2.4965413  1.4645283  1.1193455  1.0887814  0.24315852 0.42903015
 0.40263683 0.7568799  0.62619835 0.6405027  0.7195643 ]
```

## System information

**Please complete the following information:**
-  BoTorch: 0.1.3
-  GPyTorch: 0.3.5
-  PyTorch: 1.2.0
-  OS: Ubuntu 18.04.3 LTS (Bionic Beaver)

[debug_autograd.zip](https://github.com/pytorch/botorch/files/3668482/debug_autograd.zip)

**Comment by Balandat** [comment]
This seems likely to be the same issue as #279, which is caused by a bug in pytorch’s cdist function. Can you try running this on the latest gpytorch master (which avoids using torch.cdist)? Alternatively, you can use torch nightlies (since this bug has been fixed on torch master).

**Comment by shalijiang** [comment]
ha, seems it's only a problem when we pass in a single point, I figured the fastest (and nasty) way to work around this is to stack `x` into a batch or two `xx = torch.stack((x, x))`, then evaluate, only only use one of the values ...

**Comment by shalijiang** [comment]
also, tried using latest gpytorch, can fix the problem, but torch nightly did not.

**Comment by Balandat** [comment]
Hmm ok thanks for the feedback on this, will double check that this has been fixed in torch.

**Comment by Balandat** [comment]
Ok so I guess the issue is more fundamentally that in the `torch.cidst` code path we previously squared the sqrt dist - which is a null-op, but not to autograd, and sqrt is actually non-differentiable at zero. This was fixed in https://github.com/cornellius-gp/gpytorch/pull/869 by avoiding using cdist. 

I requested in that `cdist` get an option for returning the squared distances in https://github.com/pytorch/pytorch/pull/25799
